lib/leanplum_api/api.rb

```diff
@@ -228,10 +232,9 @@ def build_user_attributes_hash(user_data)
     # @param [Hash] device_data device attributes to set into LP device
     def build_device_attributes_hash(device_data)
       device_hash = fix_iso8601(device_data)
-      extract_user_id_or_device_id_hash!(device_hash).merge(
-        action: SET_DEVICE_ATTRIBUTES,
-        deviceAttributes: device_hash
-      )
+      device_hash[:action] = SET_DEVICE_ATTRIBUTES
+      # device attributes must be at top level
+      extract_user_id_or_device_id_hash!(device_hash).merge(device_hash)
```

Comment:
This line reads strange since you're both passing in `device_hash` to the method and merging it into the results.